### ISSUE TITLE
TFIDFVectorizer — generate TF-IDF representations from text data

### ISSUE THREAD

**Issue by DeathSurfing** [featrs/featrs#77]
## Description

`TFIDFVectorizer` converts text documents into a **TF-IDF** (term frequency - inverse document frequency) feature matrix. It builds on the bag-of-words counts from `CountVectorizer` (#30), then transforms counts into **TF-IDF scores** that downweight terms appearing in many documents (COMMON, NON-DISCRIMINATIVE) and emphasize distinctive terms.

Mathematically: `tfidf(t, d) = tf(t, d) * idf(t)`, where `tf` is the term frequency (raw count or normalized) and `idf(t) = ln((1 + N) / (1 + df(t))) + 1` (smoothed inverse document frequency, scikit-learn default).

This is the equivalent of scikit-learn's [`TfidfVectorizer`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html).

## Proposed API

```rust
pub enum TfWeighting {
    RawCount,            // term count (integer)
    TermFrequency,       // count / total_terms_in_doc
    Binary,              // 1 if present, else 0
    LogTf,               // 1 + ln(count)
}

pub enum IDFWeighting {
    Unsmoothed,           // ln(N / df(t))
    Smoothed,             // ln((1 + N) / (1 + df(t))) + 1  (default, sklearn)
    Probabilistic,        // ln((N - df(t) + 0.5) / (df(t) + 0.5))
}

pub enum NormMethod {
    None,
    L1,
    L2,
}

pub struct TFIDFVectorizer {
    fitted: bool,
    column: Option<String>,
    tokenizer: Tokenizer,                // reuse from #30 CountVectorizer
    ngram_range: (usize, usize),
    max_features: Option<usize>,
    min_df: usize,
    max_df: f64,
    stop_words: Option<Vec<String>>,
    lowercase: bool,
    tf_weighting: TfWeighting,            // default: TermFrequency (sklearn uses raw count by default; either fine)
    idf_weighting: IDFWeighting,           // default: Smoothed
    norm: NormMethod,                      // default: L2
    sublinear_tf: bool,                    // apply 1 + ln(count) for tf, default: false
    vocabulary: Option<HashMap<String, u32>>,
    idf: Option<Vec<f64>>,                // per-term idf, learned from training
}

impl TFIDFVectorizer {
    pub fn new() -> Self;
    pub fn column(mut self, c: &str) -> Self;
    pub fn ngram_range(mut self, lo: usize, hi: usize) -> Self;
    pub fn max_features(mut self, n: usize) -> Self;
    pub fn stop_words(mut self, words: Vec<String>) -> Self;
    pub fn lowercase(mut self, b: bool) -> Self;
    pub fn norm(mut self, n: NormMethod) -> Self;
    pub fn sublinear_tf(mut self, b: bool) -> Self;
    // ... etc, mirroring CountVectorizer setters
}

impl Default for TFIDFVectorizer { ... }

impl Fit<DataFrame, Output=()> for TFIDFVectorizer { ... }
impl Transform<DataFrame, Output=DataFrame> for TFIDFVectorizer { ... }
```

## Implementation Guidance

1. Create `src/preprocessing/text/tfidf_vectorizer.rs` (same `text/` submodule as #30)
2. **Reuse**: Can either delegate to `CountVectorizer` internally (preferred — DRY principle) or re-implement the counting logic. 
   - **Option A** (preferred): Compose. Hold a `CountVectorizer` field, call its `fit` to get vocabulary + counts, then compute IDF over the learned DF.
   - **Option B**: Re-implement tokenizer + vocabulary inline. More code, less DRY. Don't do this.
3. In `fit`:
   - Compose a `CountVectorizer` with shared parameters (`ngram_range`, `max_features`, `min_df`, `max_df`, `stop_words`, `lowercase`, tokenizer)
   - Call `count_vectorizer.fit(x)`
   - Extract the vocabulary AND document frequency (DF) — may need to extend CountVectorizer to expose DF internally (would add a `document_frequency() -> &[u32]` accessor to `CountVectorizer` — minor addition)
   - For each vocabulary term `t`: compute `idf[t]` per configured method
   - Store `idf` vector
   - Set `fitted = true`
4. In `transform`:
   - Compute count vector for each doc (via CountVectorizer or by reusing methodology)
   - Apply TF weighting (`RawCount` keeps counts; `TermFrequency` divides by doc length; `Binary` is 0/1)
   - If `sublinear_tf = true`: replace count `c` with `1 + ln(c)`
   - Multiply by `idf[t]`: `tfidf[t] = tf * idf[t]`
   - Apply normalization (L1: divide by L1 norm; L2: divide by L2 norm)
   - Output dtype: Float64
5. Register module, add to prelude

## Edge Cases

- **Empty document**: tf=0 → tfidf for all terms = 0; normalization → all zeros (handle divide-by-zero gracefully)
- **Single term in vocab**: idf computation degenerates — `smoothed: ln(2/2) + 1 = 1` (constant)
- **All documents same**: idf roughly constant; tf variation is the only signal
- **Sublinear_tf with count=0**: leave at 0 (don't apply log)
- **TF weighting Binary**: counts become 0/1, then multiplied byidf
- **Norm = None**: no normalization; values can be large
- **Norm = L1**: rows sum to 1
- **Norm = L2**: rows have unit Euclidean norm
- **Max_df=1.0**: no IDF filtering
- **Unknown term at transform**: skipped (consistent with CountVectorizer)
- **NaN propagation**: handle with care — `idf` should always be finite

## Testing

- Manual TF-IDF on 2-3 document example matches sklearn output (use approx with relative epsilon)
- IDF for term in all docs: Smoothed → `ln((1+N)/(1+N)) + 1 = 1`
- IDF for term in only 1 doc: `ln((1+N)/2) + 1`
- L2 norm: each row vector has unit length
- L1 norm: each row vector sums to 1
- Sublinear TF: count `5` → `1 + ln(5)`
- Binary TF + IDF: counts collapse to 0/1
- Not-fitted error
- Empty doc → all-zero row, no NaN
- Verifies output is close to sklearn reference for a known example

## Dependencies

- Same as #30 (`regex` for tokenizer)
- **Reuse** `CountVectorizer` — no new deps if recomposition chosen

## Related Issues

- #30 — `CountVectorizer` (direct prerequisite — depends on this for vocabulary + counting)
- #32 — `CharacterNGramVectorizer` (sibling)
- #38 — FeatureNamesOut
- #26 — `ahash` for FeatureHasher (could produce fixed-size hashed TF-IDF — reduces memory)
- #25 — Rayon parallelization (rows are independent — trivially parallelizable)
